### PR TITLE
[IMP] mass_mailing: add a mobile preview to mailing editor

### PR DESCRIPTION
**Pull request by pko-odoo**
PURPOSE

With this commit you can now see how the mailing you're creating
will look like on a mobile.

LINKS

Task-2818037

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 1]
```diff
@@ -3,13 +3,62 @@ odoo.define('mass_mailing.FieldHtml', function (require) {
...
+    init(parent, options) {
+        this._super(parent, options);
+        this.preview = options.preview;
+    },
+
+    willStart() {
+        const service = this._rpc({
+            route: '/mailing/preview/get_styling',
+        }).then((style) => {
+            if (style) {
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:29]: Actually, could style be void ?

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 2]
```diff
@@ -306,6 +355,23 @@ var MassMailingFieldHtml = FieldHtml.extend({
...
+        // disable all links to avoid landing on a black screen
+        html.querySelectorAll('a').forEach(el => {
+            el.style.setProperty('pointer-events', 'none');
+        });
+
+        if (this.mobilePreview && !this.mobilePreview.isDestroyed()) {
+            return this.mobilePreview.close();
+        }
+        this.mobilePreview = new MassMailingMobilePreviewDialog(this, {
+            title: Markup(_.escape(_t('Mobile preview')) + ' <span class="fa fa-refresh"/>'),
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:370]: Why markup + escape ? Is there any dynamic input we want to escape ?

**Review by tde-banana-odoo — COMMENTED** [review]
Bloup, some comments while going through, not much to say on my side :)

**Comment by tde-banana-odoo** [comment]
@Zinston I suppose you are already aware of this one. As it impacts the owlification, would you prefer speeding-it up, merge it and rebase the owlification on it (and test the complete job), or the inverse ?

Cheers !

**Comment by Zinston** [comment]
@tde-banana-odoo Thanks for the ping. @Goaman is the one handling the owlification (it's in its review phase right now). Not that it impacts me personally but imo it would make more sense to merge owl first (provided that the review doesn't take too long) since it's a bigger, more complex and more prioritary task. Anything we can do to prevent delaying it further. Also this way @LNAodoo can indeed test this branch in its full new context when it's done.
Please ping me when it reaches the review stage so I can also take a look, though I think @pko-odoo and I have already discussed the things on which I had my word to say, but you never know :)
cc @dmo-odoo

**Comment by Goaman** [comment]
I agree with @Zinston, merging the owl branch before seems to make more sense.

**Comment by tde-banana-odoo** [comment]
Task is waiting for your merge, don't worry :) just ping us when it is done so that we rebase and move forward, as this is a target for the freeze ^^ (aka tomorrow, no stress).

File: addons/mass_mailing/static/src/scss/mass_mailing_mobile_preview.scss [hunk 3]
```diff
@@ -0,0 +1,64 @@
+.oe_mobile_preview {
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/static/src/scss/mass_mailing_mobile_preview.scss:1]: ``o_mailing_mobile_preview`` ?

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js [hunk 4]
```diff
@@ -212,6 +214,17 @@ export class MassMailingHtmlField extends HtmlField {
             this.onIframeUpdated();
         });
 
+        $snippetsSideBar.on('click', '.o_mobile_preview_btn', () => {
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_html_field.js:217]: Can't we use a t-on-click ?

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js [hunk 5]
```diff
@@ -212,6 +214,17 @@ export class MassMailingHtmlField extends HtmlField {
...
+            let mailingHtml = new DOMParser().parseFromString(this.wysiwyg.getValue(), 'text/html');
+            [...mailingHtml.querySelectorAll('a')].forEach(el => {
+                el.style.setProperty('pointer-events', 'none');
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_html_field.js:220]: What does it do precisely ? Removing hover and stuff on links ?

File: addons/mass_mailing/static/src/js/mass_mailing_mobile_preview.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                '</html>';
+    }
+}
+
+MassMailingMobilePreviewDialog.props = {
+    ...Dialog.props,
+    preview: { type: String },
+    close: Function,
+};
+delete MassMailingMobilePreviewDialog.props.slots;
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_mobile_preview.js:43]: What does it do ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing !

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js (hunk 4 above)
Comment (jeh-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js:217]: Seems tricky or not feasible at the moment, this button is part of `snippets.editor` which is not in owl. I've done the same thing as the `o_codeview_btn` so I suppose it's at least ok for now.

File: addons/mass_mailing/static/src/js/mass_mailing_mobile_preview.js (hunk 6 above)
Comment (jeh-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_mobile_preview.js:43]: It removes the slots property that's required by default and that we do not used it here. I can remove these line but then I need to add at the creation of dialog the param: `slots: { default: {} }`
Seems it was also done for `ProductImageViewer.props.slots`

**Review by jeh-odoo — COMMENTED** [review]
Hello @tde-banana-odoo 
Here is some answers 😄 
Cheers!

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Comment by tde-banana-odoo** [comment]
@robodoo retry

**Pull request closed without merging**